core/include/detray/geometry/barcode.hpp

```diff
@@ -1,15 +1,217 @@
+    constexpr barcode& set_id(surface_id id) {
+        return set_bits(k_id_mask, static_cast<value_t>(id));
+    }
+
+    /// Set the surface index.
+    DETRAY_HOST_DEVICE
+    constexpr barcode& set_index(value_t index) {
+        return set_bits(k_index_mask, index);
+    }
+
+    /// Set the extra identifier.
+    DETRAY_HOST_DEVICE
+    constexpr barcode& set_extra(value_t extra) {
+        return set_bits(k_extra_mask, extra);
+    }
+
+    /// Check wether the barcode is valid to use.
+    DETRAY_HOST_DEVICE
+    constexpr bool is_invalid() const {
+        return ((volume() == (1UL << 12) - 1UL) |
+                (static_cast<value_t>(id()) == (1UL << 4) - 1UL) |
+                (index() == (1UL << 40) - 1UL));
+    }
+
+    private:
+    // clang-format off
+    static constexpr value_t k_volume_mask = 0xfff0000000000000; // (2^12)-1 = 4095 volumes 
+    static constexpr value_t k_id_mask     = 0x000f000000000000; // (2^4)-1 = 15 surface categories 
+    static constexpr value_t k_index_mask  = 0x0000ffffffffff00; // (2^40)-1 = 1099511627775 surfaces (per volume and id)
+    static constexpr value_t k_extra_mask  = 0x00000000000000ff; // (2^8)-1 = 255 extra values
+    // clang-format on
+
+    // Default: All bits set to 1
+    value_t m_value{~static_cast<value_t>(0)};
+
+// Use compiler builtin to extract the number of trailing bits from the
+// mask. The builtin should be available on all supported compilers.
+// Need unsigned long long version (...ll) to ensure uint64_t
+// compatibility.
+#if defined(__CUDACC__)
+    /// Extract the bit shift necessary to access the masked values - CUDA
+    ///
+    /// @note undefined behaviour for mask == 0 which we should not have.
+    DETRAY_DEVICE
+    static constexpr int extractShift(value_t mask) {
+        if (mask == k_volume_mask) {
+            return 52;
+        } else if (mask == k_id_mask) {
+            return 48;
+        } else if (mask == k_index_mask) {
+            return 8;
+        } else if (mask == k_extra_mask) {
+            return 0;
+        } else {
+            assert(false && "invalid mask");
+            return -1;
+        }
+        // A '__ctz' builtin is not available in CUDA, so the mask needs to
+        // be reversed and then the leading zeores are counted
+        // return __clzll(__brevll(mask));
+    }
+#elif !defined(__CUDACC__)
+    /// Extract the bit shift necessary to access the masked values - host/SYCL
+    ///
+    /// @note undefined behaviour for mask == 0 which we should not have.
+    DETRAY_HOST
+    static constexpr int extractShift(value_t mask) {
+        // should also be supported in SYCL:
+        // https://github.com/intel/llvm/blob/sycl/clang/docs/LanguageExtensions.rst#id91
+        return __builtin_ctzll(mask);
+    }
+#endif
+
+    /// Extract the masked bits from the encoded value.
+    DETRAY_HOST_DEVICE
+    constexpr value_t getBits(value_t mask) const {
+        return (m_value & mask) >> extractShift(mask);
```

Comment:
How frequently do you expect this method and `set_bits` to be called? If it is absolutely performance-critical, maybe it is worth considering making the `extractShift` a compile time function, such that calls to this method would become, for example:

```cpp
constexpr value_t index() const {
    return (m_value & k_index_mask) >> shift<k_index_mask>::value;
}
```

Right now the compiler cannot do this because `mask` can be anything. If we know that it will only ever be the 4 masks declared above we can template and resolve the shift at compile time.

Reply:
Yes, true. It also frightened me a little not to use ```if constexpr``` in the CUDA specific code. I want to try this bitwise encoding also at one other spot where I have very differently sized indices plus an overall memory constraint, so let me see if I can come up with something more generic

Reply:
Just to be clear this is a case where the compiler can maybe possibly resolve this at compile time, but we might want to make sure it does (in lieu of `consteval`).

Reply:
Done